core/trie/src/lib.rs

```diff
@@ -317,6 +315,138 @@ fn branch_node(has_value: bool, has_children: impl Iterator<Item = bool>) -> [u8
 	[first, (bitmap % 256 ) as u8, (bitmap / 256 ) as u8]
 }
 
+// TODO EMCH issue to add default value to HashDB trait: avoiding this costy calculation here
+// returning &'a or &'static is tricky in this one (maybe the compiler can accept both).
+// Maybe a const in Hasher trait is right (there is already a const).
+// Otherwhise store it in keyspacedb
+pub struct KeySpacedDB<'a, DB, H: Hasher>(&'a DB, &'a KeySpace, H::Out);
+pub struct KeySpacedDBMut<'a, DB, H: Hasher>(&'a mut DB, &'a KeySpace, H::Out);
+// TODO rem in favor of using underlying Memorydb values
+const NULL_NODE: &[u8] = &[0];
+impl<'a, DB, H> KeySpacedDB<'a, DB, H> where
+	H: Hasher,
+{
+	/// instantiate new keyspaced db
+	pub fn new(db: &'a DB, ks: &'a KeySpace) -> Self {
+		// TODO remove that it is already defined and probably stored in underlying db
+		let null_node_data = H::hash(NULL_NODE);
+		KeySpacedDB(db, ks, null_node_data)
+	}
+}
+impl<'a, DB, H> KeySpacedDBMut<'a, DB, H> where
+	H: Hasher,
+{
+	/// instantiate new keyspaced db
+	pub fn new(db: &'a mut DB, ks: &'a KeySpace) -> Self {
+		// TODO remove that it is already defined and probably stored in underlying db
+		let null_node_data = H::hash(&[0]);
+		KeySpacedDBMut(db, ks, null_node_data)
+	}
+}
+
+impl<'a, DB, H, T> hash_db::HashDBRef<H, T> for KeySpacedDB<'a, DB, H> where
+	DB: hash_db::HashDBRef<H, T>,
+	H: Hasher,
+	T: From<&'static [u8]>,
+{
+	fn get(&self, key: &H::Out, prefix: &[u8]) -> Option<T> {
+		if key == &self.2 {
+			return Some(NULL_NODE.into());
+		}
+		// TODO use interal buffer for derive key to avoid alloc?
+		let derived_prefix = keyspace_as_prefix_alloc(self.1, prefix);
+		self.0.get(key, &derived_prefix)
+	}
+
+	fn contains(&self, key: &H::Out, prefix: &[u8]) -> bool {
+		if key == &self.2 {
+			return true;
+		}
+
+		let derived_prefix = keyspace_as_prefix_alloc(self.1, prefix);
+		self.0.contains(key, &derived_prefix)
+	}
+}
+
+// TODO makes plainDB use prefix
```

Comment:
For mePlainDB is simply HashDB that does not return null node on null key, so it should use prefix as HashDB does, @soorpas can you confirm that point (it would require an issue for trie to change its api)?

Reply:
`PlainDB` is the raw db interface. The intention was that we can build `HashDB` on top of `PlainDB`. So we can have a PlainDB of `k` -> `v`, and multiple `HashDB` that does `h_n(k)` -> `v` so that it supports multiple hash algorithms.

For here, is it possible that we just don't implement `PlainDB` and `PlainDBRef`?

Reply:
Thanks for the explanation, it makes a lot more sense to me :). And implementation is indeed useless.